File: 1/sol.py
import fileinput

def replace(line):
    snums = {}
    snums["one"] = "1"
    snums["two"] = "2"
    snums["three"] = "3"
    snums["four"] = "4"
    snums["five"] = "5"
    snums["six"] = "6"
    snums["seven"] = "7"
    snums["eight"] = "8"
    snums["nine"] = "9"

    lmn = min(((i, sn) for sn in snums if (i := line.find(sn)) != -1), default = None)
    rmx = max(((i, sn) for sn in snums if (i := line.rfind(sn)) != -1), default = None)

    if lmn is None and rmx is None:
        return line
    elif lmn == rmx:
        i, sn = lmn
        return line[:i] + snums[sn] + line[i + len(sn):]
    else:
        li, lsn = lmn
        ri, rsn = rmx
        left = line[:li] + snums[lsn]
        mid = line[li + len(lsn):ri]
        right = line[ri:] + snums[rsn] + line[ri + len(rsn):]
        return left + mid + right

def calibration(lines):
    for line in lines:
        digits = [int(c) for c in line if c.isdigit()]
        yield digits[0] * 10 + digits[-1]

def part1(lines):
    return sum(calibration(lines))

def part2(lines):
    return sum(calibration(map(replace, lines)))

lines = [line[:-1] for line in fileinput.input(encoding = "utf-8")]

# p1
print(part1(lines))

# p2
print(part2(lines))

File: 1/sol.rs
use std::io::{self, BufRead};
use std::collections::HashMap;


fn parse_string_nums(s: &str, snums: &HashMap<&str, &str>) -> String {
    let mn = snums
        .keys()
        .filter_map(|snum| s.find(snum).map(|i| (i, snum)))
        .min();

    let mx = snums
        .keys()
        .filter_map(|snum| s.rfind(snum).map(|i| (i , snum)))
        .max();

    if mn.is_none() {
        s.to_string()
    } else if mx == mn {
        let (i, sn) = mn.unwrap();
        format!("{}{}{}", &s[..i], snums[sn], &s[i + sn.len()..])
    } else {
        let (li, lsn) = mn.unwrap();
        let (ri, rsn) = mx.unwrap();
        let left = format!("{}{}", &s[..li], snums[lsn]);
        let mid = if li + lsn.len() <= ri{
            format!("{}", &s[li + lsn.len()..ri])
        } else {
            "".to_string()
        };
        let right = format!("{}{}{}", &s[ri..], snums[rsn], &s[ri + rsn.len()..]);
        format!("{}{}{}", left, mid, right)
    }
}

fn calibration(line: &str) -> usize {
    let digits = line
        .chars()
        .filter(|c| c.is_digit(10))
        .collect::<Vec<_>>();
    let fst = digits.first().unwrap().to_digit(10).unwrap() as usize;
    let snd = digits.last().unwrap().to_digit(10).unwrap() as usize;
    return fst * 10 + snd;
}

fn main() -> io::Result<()> {
    let stdin = io::stdin();
    let lines = stdin
        .lock()
        .lines()
        .filter_map(|res_s| res_s.ok())
        .collect::<Vec<_>>();

    let mut snums = HashMap::new();
    snums.insert("one",   "1");
    snums.insert("two",   "2");
    snums.insert("three", "3");
    snums.insert("four",  "4");
    snums.insert("five",  "5");
    snums.insert("six",   "6");
    snums.insert("seven", "7");
    snums.insert("eight", "8");
    snums.insert("nine",  "9");

    // p1
    let p1 = lines
        .iter()
        .map(|s| calibration(s))
        .sum::<usize>();
    println!("{:?}", p1);

    // p2
    let p2 = lines
        .iter()
        .map(|s| calibration(&parse_string_nums(s, &snums)))
        .sum::<usize>();
    println!("{:?}", p2);
    
    Ok(())
}
